### ISSUE TITLE
Should we add constraints to radiogroup children...?

### ISSUE THREAD

**Issue by spectranaut** [w3c/aria#2910]
The role `radiogroup` has no required or allowed accessibility children, so in effect, there could be one radiogroup on the root node and all role=radios contained within are considered part of that same radiogroup.

Should we have some limits on what is contained in a radiogroup?

Right now chrome only looks for a direct parent of the aria radio to find the radiogroup, which does not match the spec, so I'm updating that. But how far up should we look?

**Comment by pkra** [comment]
> Should we have some limits on what is contained in a radiogroup?

This discussion came up in https://github.com/w3c/aria/issues/1721 which led to the note added in https://github.com/w3c/aria/pull/2159.

As I recall, there were clear arguments against limiting children while acknowledging that that can easily cause UX issues (incl. WCAG failures).

> But how far up should we look?

The spec for `radio` currently reads:

> Authors SHOULD ensure that elements with role radio are explicitly grouped in order to indicate which ones affect the same value. This is achieved by enclosing the radio elements in an element with role radiogroup. If it is not possible to make the radio buttons DOM children of the radiogroup, authors SHOULD use the aria-owns attribute on the radiogroup element to indicate the relationship to its children. 

Could it be sufficient to strengthen this to a MUST? (Perhaps allowing "intervening generics"?)

**Comment by pkra** [comment]
See also https://github.com/w3c/aria/issues/2058

**Comment by pkra** [comment]
Also https://github.com/w3c/aria/issues/2685

**Comment by spectranaut** [comment]
Wow thank you for all the links Peter! I see we return to this issue quite a bit in different ways... 

But if we aren't close to making up our mind, I wonder what the other browsers do about this particular issue, grouping radio buttons in a radiogroup -- do you search up the tree forever or do you have a heuristic, @twilco? @jcsteh?

**Comment by scottaohara** [comment]
@spectranaut i recall talking to Aaron about how chrome does this, as it was a change they made to how related radios were calculated within the last few years.  it definitely didn't used to work this way, and has definitely resulted in some regressions for properly associating radio buttons that might not be in a structure that lends itself well to chrome's assumptions.  

in HTML one _should_ be able to identify the related radios by use of the `name` attribute. i'm not sure how well this works in chrome now, if at all, with the change that was made.  and per https://github.com/w3c/aria/issues/1721, making an ARIA specific way to emulate the `name` attribute association was not agreed upon.

**Comment by twilco** [comment]
> do you search up the tree forever or do you have a heuristic

WebKit will search the ancestry up to the root for a `radiogroup`, and then collect all radio buttons within that entire subtree:

https://github.com/WebKit/WebKit/blob/fa43159ee89ca8e328270d9a40fb009748efbf82/Source/WebCore/accessibility/AXCoreObject.cpp#L2060

Beyond the original issue being discussed, I'm realizing now that this would also collect radio buttons that belong to some other group (e.g. a non-ARIA radio with a `name` association). Filed a bug for myself here: https://bugs.webkit.org/show_bug.cgi?id=325051

**Comment by jcsteh** [comment]
For ARIA radio buttons, Gecko only considers siblings, skipping intervening generics during the walk. So there can be any number of intervening generics between the radios, whether those are direct siblings or radios nested in generics, and they will be ignored for the purposes of grouping. However, any non-generic element will break the group. Gecko doesn't actually even look at whether there's a radiogroup in the ancestry at all.

HTML `<input type="radio">` is handled differently. For those, we group purely based on the name attribute. We have some bugs there regarding shadow DOM scoping, though.

**Comment by jcsteh** [comment]
Just a correction to the minutes regarding Firefox:

> Firefox goes up until it sees a potential conflict.

Firefox doesn't go up at all. As noted in https://github.com/w3c/aria/issues/2910#issuecomment-5808683020, it walks siblings, flattening generics. It doesn't care if there is a radiogroup parent.
